### **6. Разбить данные на обучающую и тестовую выборки**
### **7. Запустить классификатор (ререссию) ближайших соседей или другой метод (аргументируйте выбор!)**
### **8. Вычислить ошибки на обучающей и тестовой выборках. Выбрать оптимальное значение гиперпараметра (к-ва ближайших соседей) Сделать выводы**

In [3]:
# Подключение к моему гугл диску, чтобы можно было обратиться к датасету напрямую.
from google.colab import drive
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [4]:
import pandas as pd
dataset = pd.read_csv('/content/gdrive/MyDrive/Machine_Learning_Ermolaev_382003_3/After_3-5_Horror_movie_IMDB.csv')
dataset.to_excel('subset_data.xlsx', index=False)

dataset

Title  Release Date  Release Country  \
0                               Gut (2012)           0.0         0.929577   
1          The Haunting of Mia Moss (2017)           1.0         0.929577   
2                      Sleepwalking (2017)           1.0         0.140845   
3      Treasure Chest of Horrors II (2013)           0.2         0.929577   
4                           Infidus (2015)           0.6         0.929577   
...                                    ...           ...              ...   
3298            Victor Frankenstein (2015)           0.6         0.929577   
3299  The Exorcism of Molly Hartley (2015)           0.6         0.929577   
3300                    Talon Falls (2017)           1.0         0.929577   
3301    BloodRayne: The Third Reich (2011)           0.0         0.436620   
3302                     13 Cameras (2015)           0.8         0.929577   

      Movie Rating  Review Rating  Movie Run Time  \
0         0.392667       3.900000        0.470238   
1         0.392667       3.833333        0.452381   
2         0.392667       3.766667        0.434524   
3         0.200000       3.700000        0.416667   
4         0.200000       5.800000        0.404762   
...            ...            ...             ...   
3298      0.400000       6.000000        0.583333   
3299      0.900000       4.600000        0.500000   
3300      0.700000       4.600000        0.375000   
3301      0.500000       3.000000        0.398810   
3302      0.500000       5.100000        0.446429   

                                                   Plot      Cast  \
0     Directed by Elias. With Jason Vail, Nicholas W...  0.422878   
1     Directed by Jake Zelch. With Nicola Fiore, Bri...  0.725586   
2     Directed by David Briggs. With Alysia Topol, A...  0.046547   
3     Directed by M. Kelley, Shawn C. Phillips, Alex...  0.967448   
4     Directed by Giulio De Santi. With Bonini Mino,...  0.128080   
...                                                 ...       ...   
3298  Directed by Paul McGuigan. With Daniel Radclif...  0.226650   
3299  Directed by Steven R. Monroe. With Sarah Lind,...  0.839367   
3300  Directed by Joshua Shreve. With Brad Bell, Jor...  0.129297   
3301  Directed by Uwe Boll. With Natassia Malthe, Br...  0.713721   
3302  Directed by Victor Zarcoff. With Neville Archa...  0.718588   

      Filming Locations    Budget  ...  Romanian  Latin  Serbian  Catalan  \
0              0.897436  0.018088  ...         0      0        0        0   
1              0.935897  0.000158  ...         0      0        0        0   
2              0.153846  0.004592  ...         0      0        0        0   
3              0.897436  0.009026  ...         0      0        0        0   
4              0.935897  0.013461  ...         0      0        0        0   
...                 ...       ...  ...       ...    ...      ...      ...   
3298           0.884615  0.210526  ...         0      0        0        0   
3299           0.153846  0.157895  ...         0      0        0        0   
3300           0.935897  0.105263  ...         0      0        0        0   
3301           0.217949  0.052632  ...         0      0        0        0   
3302           0.897436  0.052632  ...         0      0        0        0   

      Ukrainian  Persian  American Sign Language  Bulgarian  Polish  Bengali  
0             0        0                       0          0       0        0  
1             0        0                       0          0       0        0  
2             0        0                       0          0       0        0  
3             0        0                       0          0       0        0  
4             0        0                       0          0       0        0  
...         ...      ...                     ...        ...     ...      ...  
3298          0        0                       0          0       0        0  
3299          0        0                       0          0       0        0  
3300      

### **6. Разбить данные на обучающую и тестовую выборки**

In [5]:
print(dataset.columns)

Index(['Title', 'Release Date', 'Release Country', 'Movie Rating',
       'Review Rating', 'Movie Run Time', 'Plot', 'Cast', 'Filming Locations',
       'Budget', ' Action', ' Adult', ' Adventure', ' Animation', ' Biography',
       ' Comedy', ' Crime', ' Drama', ' Family', ' Fantasy', ' History',
       ' Horror', ' Music', ' Musical', ' Mystery', ' Reality-TV', ' Romance',
       ' Sci-Fi', ' Sport', ' Thriller', ' War', ' Western', 'English',
       'Spanish', 'Japanese', 'German', 'Hindi', 'Filipino', 'French',
       'Tagalog', 'Thai', 'Tamil', 'Turkish', 'Russian', 'Indonesian',
       'Italian', 'Mandarin', 'Korean', 'Swedish', 'Portuguese', 'Telugu',
       'Chinese', 'Malay', 'Malayalam', 'Cantonese', 'Kannada', 'Norwegian',
       'Danish', 'Dutch', 'Arabic', 'Hungarian', 'Vietnamese', 'Romanian',
       'Latin', 'Serbian', 'Catalan', 'Ukrainian', 'Persian',
       'American Sign Language', 'Bulgarian', 'Polish', 'Bengali'],
      dtype='object')


In [6]:
from sklearn.model_selection import train_test_split
# Определение признаков (features) и целевой переменной (target)
#selected_features = ['Release Date','Release Country','Movie Rating',
#            'Movie Run Time','Cast','Filming Locations','Budget']
X = dataset.drop(['Title','Review Rating', 'Plot'], axis=1)  # Признаки (все столбцы, кроме 'Title','Review Rating','Plot')
y = dataset['Review Rating']  # Целевая переменная

# Разделение на тренировочную и тестовую выборки (например, 80% для обучения, 20% для тестирования)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

N_train, _ = X_train.shape
N_test,  _ = X_test.shape

N_train, N_test

(2642, 661)



---


### **Запустить классификатор (ререссию) ближайших соседей или другой метод (аргументируйте выбор!)**

**1. K-ближайших соседей (KNN):**

*Когда использовать:* KNN хорошо работает в задачах, где объекты с похожими признаками обычно имеют схожие целевые переменные.

*Преимущества:* Прост в реализации, не требует предположений о данных.

*Недостатки:* Медленный при большом количестве данных, чувствителен к шуму и выбросам.

**2. Линейная регрессия:**

*Когда использовать:* Хорошо подходит, когда существует линейная зависимость между признаками и целевой переменной.

*Преимущества*: Прост в интерпретации, быстрый, подходит для больших данных.

*Недостатки:* Предполагает линейность, не учитывает сложные взаимосвязи.

**3. Случайный лес (Random Forest):**

*Когда использовать:* Эффективен в задачах с большим количеством данных и высокой размерностью признаков. Дает хорошие результаты без настройки гиперпараметров.

*Преимущества:* Меньше подвержен переобучению, хорошо работает на больших наборах данных, учитывает важность признаков.

*Недостатки:* Может быть медленным при огромных данных, сложно интерпретировать.

**4. Нейронные сети (MLPRegressor в scikit-learn):**

*Когда использовать:* Подходит для задач с большими объемами данных и сложными взаимосвязями между признаками.

*Преимущества:* Может моделировать сложные зависимости, способен извлекать паттерны из данных.

*Недостатки:* Требуется настройка гиперпараметров, большой объем данных для обучения, может переобучаться.



---


***Исходя из приведённых мною данных, я буду использовать классификатор "Случайный лес" и "Нейронные сети"***. Свой выбор я обосновываю тем, что в прошлых заданиях мы не нашли сильной связи между признаками, при этом признаков стало достаточно много.

**Случайный Лес**:
Сначала просто реализуем данный способ классификации и посмотрим на результаты среднеквадратичной ошибки

In [7]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score, mean_squared_error

# Разделение данных на обучающий и тестовый наборы
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Инициализация модели случайного леса
random_forest = RandomForestRegressor(n_estimators=100, random_state=42)

# Обучение модели
random_forest.fit(X_train, y_train)

# Предсказание на тестовой выборке
predictions_test = random_forest.predict(X_test)

# Предсказание на тренировочной выборке (для целей сравнения)
predictions_train = random_forest.predict(X_train)

# Оценка точности модели на тестовой выборке (с использованием R-квадрата и MSE)
r2_test = r2_score(y_test, predictions_test)
mse_test = mean_squared_error(y_test, predictions_test)

# Оценка точности модели на тренировочной выборке (для сравнения)
r2_train = r2_score(y_train, predictions_train)
mse_train = mean_squared_error(y_train, predictions_train)

# Вывод результатов
print("Тестовая выборка: R-квадрат =", r2_test, "MSE =", mse_test)
print("Тренировочная выборка: R-квадрат =", r2_train, "MSE =", mse_train)

Тестовая выборка: R-квадрат = 0.06259504908193925 MSE = 1.84142349033451
Тренировочная выборка: R-квадрат = 0.8751148731062484 MSE = 0.2652175547459837


**Тестовая выборка:**

**R-квадрат = 0.06259504908193925**
Это означает, что наша модель на тестовой выборке объясняет около 6.26% вариации зависимой переменной. Это достаточно низкий показатель и может указывать на то, что ваша модель не очень хорошо подходит для предсказания тестовых данных.

**MSE = 1.84142349033451**
Это средняя ошибка квадрата между фактическими и предсказанными значениями. Чем меньше значение MSE, тем лучше. Значение около 1.84 означает, что модель имеет некоторую ошибку в среднем при предсказании на тестовых данных.

**Тренировочная выборка:**

**R-квадрат = 0.8751148731062484**
Это означает, что ваша модель на тренировочной выборке объясняет около 87.51% вариации зависимой переменной. Это высокий показатель и говорит о хорошем соответствии модели тренировочным данным.

**MSE = 0.2652175547459837**
Значение около 0.27 показывает, что модель имеет меньшую ошибку при предсказании на тренировочных данных по сравнению с тестовыми.

### ***Обратим внимание, что большая разница между показателями на тестовой и тренировочной выборках может свидетельствовать о переобучении модели.***



### **Подберём гиперпараметры.**

In [12]:
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import r2_score, mean_squared_error

# Задаем модель
model = RandomForestRegressor()

# Задаем сетку параметров для перебора
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

# Задаем Grid Search с кросс-валидацией
grid_search = GridSearchCV(model, param_grid, cv=2, scoring='neg_mean_squared_error', n_jobs=1, verbose=2)

grid_search.fit(X, y)

# Получаем лучшие параметры
best_params = grid_search.best_params_

# Получаем лучшую модель
best_model = grid_search.best_estimator_

# Получим предсказания на тестовой и тренировочной выборках
predictions_test = best_model.predict(X_test)
predictions_train = best_model.predict(X_train)

# Рассчитаем R-квадрат для тестовой и тренировочной выборок
r2_test = r2_score(y_test, predictions_test)
r2_train = r2_score(y_train, predictions_train)

# Рассчитаем среднеквадратичную ошибку для тестовой и тренировочной выборок
mse_test = mean_squared_error(y_test, predictions_test)
mse_train = mean_squared_error(y_train, predictions_train)

# Выведем результаты
print("Тестовая выборка: R-квадрат =", r2_test, "MSE =", mse_test)
print("Тренировочная выборка: R-квадрат =", r2_train, "MSE =", mse_train)

Fitting 2 folds for each of 108 candidates, totalling 216 fits
[CV] END max_depth=None, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   2.0s
[CV] END max_depth=None, min_samples_leaf=1, min_samples_split=2, n_estimators=100; total time=   2.1s
[CV] END max_depth=None, min_samples_leaf=1, min_samples_split=2, n_estimators=200; total time=   2.4s
[CV] END max_depth=None, min_samples_leaf=1, min_samples_split=2, n_estimators=200; total time=   2.4s
[CV] END max_depth=None, min_samples_leaf=1, min_samples_split=2, n_estimators=300; total time=   3.8s
[CV] END max_depth=None, min_samples_leaf=1, min_samples_split=2, n_estimators=300; total time=   4.9s
[CV] END max_depth=None, min_samples_leaf=1, min_samples_split=5, n_estimators=100; total time=   1.0s
[CV] END max_depth=None, min_samples_leaf=1, min_samples_split=5, n_estimators=100; total time=   1.0s
[CV] END max_depth=None, min_samples_leaf=1, min_samples_split=5, n_estimators=200; total time=   2.0s
[CV] END m

БЫЛО:

Тестовая выборка: R-квадрат = 0.06259504908193925 MSE = 1.84142349033451

Тренировочная выборка: R-квадрат = 0.8751148731062484 MSE = 0.2652175547459837


СТАЛО:

Тестовая выборка: R-квадрат = 0.4145382579581587 MSE = 1.1500718056077837

Тренировочная выборка: R-квадрат = 0.4044688398505262 MSE = 1.26472480749655

Т.к. необходимо чтоб R-квадрат быль большим, а MSE малым, мы видим, что результаты на тестовой и тренировочной выброках стали приблизительно одинаковыми. И при этом Значение R-квадрат стало куда меньше, что свидетельствует о том, что модель стала справляться куда лучше.

In [16]:
# Выберем случайный фильм из вашего датасета
selected_features = ['Release Date','Release Country','Movie Rating',
            'Movie Run Time','Cast','Filming Locations','Budget']

random_film = dataset.sample(1)
print("Название фильма:", random_film['Title'].values[0])

#random_film = random_film.loc[:, selected_features]

# Оставим только признаки, которые использовались для обучения модели (исключая 'Title', 'Review Rating', 'Plot')
selected_features = random_film.drop(['Title','Review Rating' ,'Plot'], axis=1)

# Предсказание оценки для отдельного случайного фильма
predicted_rating = best_model.predict(selected_features)

print("Предсказанная оценка фильма:", predicted_rating)

Название фильма: The Other Side of the Door (2016)
Предсказанная оценка фильма: [5.07982937]


**Нейронные сети (MLPRegressor в scikit-learn)**

In [8]:
from sklearn.neural_network import MLPRegressor
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

model2 = MLPRegressor(hidden_layer_sizes=(100,50), max_iter=1000)
model2.fit(X_train, y_train)

# Предсказание на тестовой выборке
predictions_test = model2.predict(X_test)

# Предсказание на тренировочной выборке (для целей сравнения)
predictions_train = model2.predict(X_train)

# Оценка точности модели на тестовой выборке (с использованием R-квадрата и MSE)
r2_test = r2_score(y_test, predictions_test)
mse_test = mean_squared_error(y_test, predictions_test)

# Оценка точности модели на тренировочной выборке (для сравнения)
r2_train = r2_score(y_train, predictions_train)
mse_train = mean_squared_error(y_train, predictions_train)

# Вывод результатов
print("Тестовая выборка: R-квадрат =", r2_test, "MSE =", mse_test)
print("Тренировочная выборка: R-квадрат =", r2_train, "MSE =", mse_train)


Тестовая выборка: R-квадрат = -0.5199409184038373 MSE = 2.985747950582442
Тренировочная выборка: R-квадрат = 0.4892552006224 MSE = 1.084664684733827


In [14]:
from sklearn.neural_network import MLPRegressor
from sklearn.model_selection import GridSearchCV
from sklearn.datasets import make_regression

# Создание модели
model2 = MLPRegressor()

# Задание сетки параметров для подбора
param_grid = {
    'hidden_layer_sizes': [(50,50), (50,100,50), (100,)],
    'activation': ['relu', 'tanh', 'logistic'],
    'solver': ['adam', 'sgd'],
    'alpha': [0.0001, 0.05],
    'max_iter': [1000],
}

# Поиск по сетке
grid_search = GridSearchCV(model2, param_grid, cv=2, scoring='neg_mean_squared_error', n_jobs=1, verbose=2)
grid_search.fit(X, y)

# Получаем лучшие параметры
best_params = grid_search.best_params_

# Получаем лучшую модель
best_model = grid_search.best_estimator_

# Предсказание на тестовой выборке
predictions_test = best_model.predict(X_test)

# Предсказание на тренировочной выборке (для целей сравнения)
predictions_train = best_model.predict(X_train)

# Оценка точности модели на тестовой выборке (с использованием R-квадрата и MSE)
r2_test = r2_score(y_test, predictions_test)
mse_test = mean_squared_error(y_test, predictions_test)

# Оценка точности модели на тренировочной выборке (для сравнения)
r2_train = r2_score(y_train, predictions_train)
mse_train = mean_squared_error(y_train, predictions_train)

# Вывод результатов
print("Тестовая выборка: R-квадрат =", r2_test, "MSE =", mse_test)
print("Тренировочная выборка: R-квадрат =", r2_train, "MSE =", mse_train)


Fitting 2 folds for each of 36 candidates, totalling 72 fits
[CV] END activation=relu, alpha=0.0001, hidden_layer_sizes=(50, 50), max_iter=1000, solver=adam; total time=  10.6s
[CV] END activation=relu, alpha=0.0001, hidden_layer_sizes=(50, 50), max_iter=1000, solver=adam; total time=   3.7s


/usr/local/lib/python3.10/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


[CV] END activation=relu, alpha=0.0001, hidden_layer_sizes=(50, 50), max_iter=1000, solver=sgd; total time=  15.0s


/usr/local/lib/python3.10/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


[CV] END activation=relu, alpha=0.0001, hidden_layer_sizes=(50, 50), max_iter=1000, solver=sgd; total time=  14.4s
[CV] END activation=relu, alpha=0.0001, hidden_layer_sizes=(50, 100, 50), max_iter=1000, solver=adam; total time=   9.6s
[CV] END activation=relu, alpha=0.0001, hidden_layer_sizes=(50, 100, 50), max_iter=1000, solver=adam; total time=   5.9s


/usr/local/lib/python3.10/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


[CV] END activation=relu, alpha=0.0001, hidden_layer_sizes=(50, 100, 50), max_iter=1000, solver=sgd; total time=  25.1s
[CV] END activation=relu, alpha=0.0001, hidden_layer_sizes=(50, 100, 50), max_iter=1000, solver=sgd; total time=  27.3s
[CV] END activation=relu, alpha=0.0001, hidden_layer_sizes=(100,), max_iter=1000, solver=adam; total time=  10.4s
[CV] END activation=relu, alpha=0.0001, hidden_layer_sizes=(100,), max_iter=1000, solver=adam; total time=   7.6s


/usr/local/lib/python3.10/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


[CV] END activation=relu, alpha=0.0001, hidden_layer_sizes=(100,), max_iter=1000, solver=sgd; total time=  13.8s
[CV] END activation=relu, alpha=0.0001, hidden_layer_sizes=(100,), max_iter=1000, solver=sgd; total time=   3.9s
[CV] END activation=relu, alpha=0.05, hidden_layer_sizes=(50, 50), max_iter=1000, solver=adam; total time=   5.9s
[CV] END activation=relu, alpha=0.05, hidden_layer_sizes=(50, 50), max_iter=1000, solver=adam; total time=   3.9s
[CV] END activation=relu, alpha=0.05, hidden_layer_sizes=(50, 50), max_iter=1000, solver=sgd; total time=  12.2s
[CV] END activation=relu, alpha=0.05, hidden_layer_sizes=(50, 50), max_iter=1000, solver=sgd; total time=   6.5s
[CV] END activation=relu, alpha=0.05, hidden_layer_sizes=(50, 100, 50), max_iter=1000, solver=adam; total time=   7.2s
[CV] END activation=relu, alpha=0.05, hidden_layer_sizes=(50, 100, 50), max_iter=1000, solver=adam; total time=   5.1s
[CV] END activation=relu, alpha=0.05, hidden_layer_sizes=(50, 100, 50), max_iter=1

/usr/local/lib/python3.10/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


[CV] END activation=relu, alpha=0.05, hidden_layer_sizes=(50, 100, 50), max_iter=1000, solver=sgd; total time=  22.6s
[CV] END activation=relu, alpha=0.05, hidden_layer_sizes=(100,), max_iter=1000, solver=adam; total time=  11.2s
[CV] END activation=relu, alpha=0.05, hidden_layer_sizes=(100,), max_iter=1000, solver=adam; total time=   7.2s


/usr/local/lib/python3.10/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


[CV] END activation=relu, alpha=0.05, hidden_layer_sizes=(100,), max_iter=1000, solver=sgd; total time=  13.2s
[CV] END activation=relu, alpha=0.05, hidden_layer_sizes=(100,), max_iter=1000, solver=sgd; total time=   7.7s
[CV] END activation=tanh, alpha=0.0001, hidden_layer_sizes=(50, 50), max_iter=1000, solver=adam; total time=  13.8s
[CV] END activation=tanh, alpha=0.0001, hidden_layer_sizes=(50, 50), max_iter=1000, solver=adam; total time=   9.2s
[CV] END activation=tanh, alpha=0.0001, hidden_layer_sizes=(50, 50), max_iter=1000, solver=sgd; total time=  11.9s
[CV] END activation=tanh, alpha=0.0001, hidden_layer_sizes=(50, 50), max_iter=1000, solver=sgd; total time=  10.3s
[CV] END activation=tanh, alpha=0.0001, hidden_layer_sizes=(50, 100, 50), max_iter=1000, solver=adam; total time=  32.6s
[CV] END activation=tanh, alpha=0.0001, hidden_layer_sizes=(50, 100, 50), max_iter=1000, solver=adam; total time=  15.3s
[CV] END activation=tanh, alpha=0.0001, hidden_layer_sizes=(50, 100, 50), 

In [19]:
from sklearn.neural_network import MLPRegressor
# Выберем случайный фильм из вашего датасета
selected_features = ['Release Date','Release Country','Movie Rating',
            'Movie Run Time','Cast','Filming Locations','Budget']

#random_film = dataset.sample(1)
random_film = dataset[dataset['Title'] == "The Other Side of the Door (2016)"]
# Оставим только признаки, которые использовались для обучения модели (исключая 'Title', 'Review Rating', 'Plot')
selected_features = random_film.drop(['Title','Review Rating' ,'Plot'], axis=1)

print("Название фильма:", random_film['Title'].values[0])
# Предсказание оценки для отдельного случайного фильма
predicted_rating = best_model.predict(selected_features)
print("Предсказанная оценка фильма:", predicted_rating)

Название фильма: The Other Side of the Door (2016)
Предсказанная оценка фильма: [4.46171901]


## ***ВЫВОД***
Я выполнил 8 пунктов из домашнего задания:

0. Описать задачу словами
1. Прочитать данные
2. Визуализировать данные и вычислить основные характеристики (среднее, разброс, корреляционную матрицу и т.д.). Интерпретировать.
3. Обработать пропущенные значения (или убедиться, что их нет)
4. Обработать категориальные признаки
5. Провести нормализацию (или объяснить, почему в вашем случае она не нужна)
6. Разбить данные на обучающую и тестовую выборки
7. Запустить классификатор (ререссию) ближайших соседей или другой метод (аргументируйте выбор!)
8. Вычислить ошибки на обучающей и тестовой выборках. Выбрать оптимальное значение гиперпараметра (к-ва ближайших соседей) Сделать выводы

После того, как я попытался решить задачу регрессии двумя разными способами, а именно при помощи метода "Случайный лес" и "Нейронные сети", я обнаружил, что результат полученный "Случайного леса" оказался лучшим.

Что могло к этому привести?

**Сложность данных:** Возможно, данные имеют более сложную или нелинейную структуру, которую легче уловить и предсказать с помощью ансамбля деревьев (случайный лес) в сравнении с нейронными сетями.

**Переобучение:** MLPRegressor (нейронные сети) могут иметь больше параметров для обучения, и могут быть более склонны к переобучению при недостаточном контроле или при работе с небольшими данными.

**Гиперпараметры:** Возможно, набор гиперпараметров, использованный в MLPRegressor, не оптимален. Нейронные сети требуют аккуратного подбора гиперпараметров, таких как количество слоев, количество нейронов в слоях, скорость обучения и другие, чтобы достичь оптимальной производительности.

Возможно необходимо дополнительно как-то иначе преобразовать изначальные текстовые признаки чтобы методы работали куда лучше, чем это есть на самом деле.